pkg/tbtc/coordination.go

```diff
@@ -0,0 +1,257 @@
+	)
+}
+
+// coordinationProposal represents a single action proposal for the given wallet.
+type coordinationProposal interface {
+	// actionType returns the specific type of the walletAction being subject
+	// of this proposal.
+	actionType() WalletActionType
+	// validityBlocks returns the number of blocks for which the proposal is
+	// valid.
+	validityBlocks() uint64
+}
+
+// noopProposal is a proposal that does not propose any action.
+type noopProposal struct{}
+
+func (np *noopProposal) actionType() WalletActionType {
+	return ActionNoop
+}
+
+func (np *noopProposal) validityBlocks() uint64 {
+	// Panic to make sure that the proposal is not processed by the node.
+	panic("noop proposal does not have validity blocks")
+}
+
+// coordinationResult represents the result of the coordination procedure
+// executed for the given wallet in the given coordination window.
+type coordinationResult struct {
+	wallet   wallet
+	window   *coordinationWindow
+	leader   chain.Address
+	proposal coordinationProposal
+	faults   []*coordinationFault
+}
+
+func (cr *coordinationResult) String() string {
+	return fmt.Sprintf(
+		"wallet [%s], window [%v], leader [%s], proposal [%s], faults [%s]",
+		&cr.wallet,
+		cr.window.coordinationBlock,
+		cr.leader,
+		cr.proposal.actionType(),
+		cr.faults,
+	)
+}
+
+// coordinationExecutor is responsible for executing the coordination
+// procedure for the given wallet.
+type coordinationExecutor struct {
+	lock *semaphore.Weighted
+
+	signers             []*signer // TODO: Do we need whole signers?
+	broadcastChannel    net.BroadcastChannel
+	membershipValidator *group.MembershipValidator
+	protocolLatch       *generator.ProtocolLatch
+}
+
+// newCoordinationExecutor creates a new coordination executor for the
+// given wallet.
+func newCoordinationExecutor(
+	signers []*signer,
+	broadcastChannel net.BroadcastChannel,
+	membershipValidator *group.MembershipValidator,
+	protocolLatch *generator.ProtocolLatch,
+) *coordinationExecutor {
+	return &coordinationExecutor{
+		lock:                semaphore.NewWeighted(1),
+		signers:             signers,
+		broadcastChannel:    broadcastChannel,
+		membershipValidator: membershipValidator,
+		protocolLatch:       protocolLatch,
+	}
+}
+
+// wallet returns the wallet this executor is responsible for.
+func (ce *coordinationExecutor) wallet() wallet {
+	// All signers belong to one wallet. Take that wallet from the
+	// first signer.
+	return ce.signers[0].wallet
+}
+
+// coordinate executes the coordination procedure for the given coordination
+// window.
+func (ce *coordinationExecutor) coordinate(
+	window *coordinationWindow,
+) (*coordinationResult, error) {
+	if lockAcquired := ce.lock.TryAcquire(1); !lockAcquired {
+		return nil, errCoordinationExecutorBusy
+	}
+	defer ce.lock.Release(1)
+
+	// TODO: Implement coordination logic. Remember about:
+	//       - Setting up the right context
+	//       - Using the protocol latch
+	//       - Using the membership validator
```

Comment:
It is something I am not yet fully following. My understanding of this function is that it is doing the work of the centralized coordinator. For example, it checks if the sweeping is necessary or if redemption is necessary and then proposes it. 

If so, it does not feel like the right place to use the protocol latch or membership validator. Shouldn't this happen later in the coordinator result processor that will fire the appropriate proposal handler?

Reply:
Sure, I'm aware that the full picture may not be completely visible yet so, let me explain this.

> My understanding of this function is that it is doing the work of the centralized coordinator. For example, it checks if the sweeping is necessary or if redemption is necessary and then proposes it.

This is true but that's just the part of the job. Here is the pseudocode of the `coordinationExecutor.coordinate` function:
```
func (ce *coordinationExecutor) coordinate(
	window *coordinationWindow,
) (*coordinationResult, error) {
    	ce.acquireLock()
        ce.lockProtocolLatch()
        
        seed := ce.coordinationSeed(window)

        ctx := withCancelOnBlock(context.Background, window.activePhaseEndBlock())

        if isLeader(seed) {
            proposal := c.leaderRoutine() // This is the please where the leader checks if sweeping is necessary, etc.
            message := ce.coordinationMessage(proposal)
            ce.broadcastChannel.Send(ctx, message)
            return ce.coordinationResult(proposal)
        } else {
            var proposalChan
            ce.broadcastChannel.Recv(ctx, func(message net.Message) {
		if ce.membershipValidator.IsValidMembership(message) {
                     proposalChan <-message.proposal
                 }
	    })
            return ce.coordinationResult(<-proposalChan)
        }
}
```

Reply:
Alright, so this is leader and follower logic. This makes sense. Would be good to make it crystal-clear in the `coordinate` function docs.